spacy/ml/models/entity_linker.py

```diff
@@ -1,25 +1,27 @@
 from pathlib import Path
-from typing import Optional, Callable, Iterable, List
+from typing import Optional, Callable, Iterable, List, Tuple
 from thinc.types import Floats2d
 from thinc.api import chain, clone, list2ragged, reduce_mean, residual
-from thinc.api import Model, Maxout, Linear
+from thinc.api import Model, Maxout, Linear, noop, tuplify, Ragged
 
 from ...util import registry
 from ...kb import KnowledgeBase, Candidate, get_candidates
 from ...vocab import Vocab
 from ...tokens import Span, Doc
+from ..extract_spans import extract_spans
+from ...errors import Errors
 
 
 @registry.architectures("spacy.EntityLinker.v1")
```

Comment:
We probably need to bump this version? (and move v1 to legacy)

Reply:
Good point. But because the pipeline code also has to change, the v1 model by itself won't work, so how should that be handled?

Reply:
Ah, right. Hm. What we can do is to keep v1 in spaCy's codebase for now and raise an exception explaining users need to move to v2 for the new code. We can't move the function and raise this error from within `spacy-legacy`, because then people wouldn't be able to update `spacy-legacy` without updating `spacy`, and we'll get entangled in version misery.

Reply:
Make sure to go through the code base and update docs & defaults to v2 too.

Reply:
To be clear, when you say there should be a stripped v1, should I retain the pipeline code to make that functional? Or should it be a non-functional stub that tells people v1 doesn't work?

Reply:
For now I created a stub v1 that throws an error and shows a short explanation and links to this PR.

Reply:
Yea I think that makes sense, I don't see a better way to do it. This probably means we need to move this PR to 3.3 (`develop`) though, because it will break people's config files.

Reply:
@adrianeboyd: do you see a better way? I don't think we've encountered this before, that we actually have to retire a previous version of a registered architecture.

Reply:
There are quite a few older architectures in `spacy-legacy` and I don't see why you can't also include the older component? It might require a lot of copying to do it cleanly (plus all the registry substitutions), but I think that would be better than removing it in a breaking way, even if it's broken/buggy.

Reply:
So the solution would be to
 
1. copy the old component code somewhere in the component (not in legacy?)
2. In the component, check if the architecture is v1 and use the old component code if so

Is that right? It feels a little inside out and it'll leave a lot of junk around but it's not too tangled.

Reply:
No, I think an entirely independent copy should be in `spacy-legacy` if there's no other way to preserve it.

Reply:
I think we have made minor changes to the parser algorithm across minor versions without the whole copying rigamarole. So I guess doing it over a minor version (also with the architecture version stuff) is the other option.

Reply:
Oof, that feels like quite the hassle, but you're right that at least it wouldn't be breaking. I'm just not sure whether it's all worth it, because the v2 is just the v1+bugfix, it's not like you're losing anything by removing it. People will just have to change the `1` in their config file to a `2`...

Reply:
So this does change a lot of code, but since the architecture isn't changing and it really is just a bugfix, could we just skip raising the version entirely? The only change in behavior from outside should be that listeners work now.

Reply:
Actually, we had to change the config for this, so that's reason enough we can't use the same version I guess. The good news (?) is that besides changing the version number, users can explicitly control where NER annotations come from now.

Reply:
Going over actually implementing this I am still kind of confused about what this would look like. Assuming the v1 component code is in spacy-legacy, how is it actually accessed? 

1. The non-legacy update/init methods check if the architecture is v1 and change their behavior, importing from legacy (simple implementation but very ugly)
2. Legacy registers a new component, `entity_linker.v1` or something, and users have to update their configs. This is a little weird since components don't usually have version numbers. There is also a question of whether the whole pipeline should be copied over, or whether it should be subclassed and have select methods overridden (probably the latter?). 

Moving the component code inside the model in spacy-legacy doesn't work because the model only receives Docs, not Examples. 

I guess option 2 here is preferable, but since this is the first time we've put a component in legacy we need to think about not just this case but future ones too.

Reply:
I wouldn't go for option 2 because if users need to update their configs then we might as well tell them to move to v2 :| 

(Honestly I'd still prefer just making v1 obsolete)

Reply:
Note that as a result of discussion, we decided to keep the v1 component in spacy core in a new legacy area. The old component will be swapped in based on the model, as even if that's messy it's not disruptive to existing users.